### ISSUE TITLE
New rule: ensure that all code paths return a value

### ISSUE THREAD

**Issue by adidahiya** [palantir/tslint#580]
tsc doesn't report any errors for code like this:

``` ts
function foo(): string {
  if (false) {
    return "abc";
  }
}
```

It'd be nice to have a tslint rule that does control flow analysis to warn against that.

**Comment by aluanhaddad** [comment]
@adidahiya 

This is an error since TypeScript 1.8. 

See https://github.com/Microsoft/TypeScript/pull/4788 and http://www.typescriptlang.org/play/#src=function%20foo()%3A%20string%20%7B%0D%0A%20%20if%20(false)%20%7B%0D%0A%20%20%20%20return%20%22abc%22%3B%0D%0A%20%20%7D%0D%0A%7D

**Comment by andy-hanson** [comment]
Can this be closed then?
